Add -r timeout[:interval] option to atf-check.

cemeyer opened the pull request:
This allows repeatedly polling some command until it succeeds, to some
timeout limit.

cemeyer commented:
Polling is absolutely not ideal; but we end up doing it in many tests anyway, and it isn't an awful fit for test code.  Might as well have a single decent implementation that can be used broadly instead of reimplementing the wheel everywhere.

asomers commented:
This is arguably beyond the scope of atf-sh.  But it might still be worthwhile if it saves enough headaches for consumers.  Do you happen to know how many places in FreeBSD could currently use this feature?

cemeyer commented:
This isn't a perfect metric (both false positives and negatives), but for some context:

```
$ find . -type f -name \*test\*.sh -exec rg 'sleep ' {} \+ | wc -l
245
```

Many (but not all) of these are poll-sleeps or broken poll-sleeps (i.e., hardcode 1 second and assume the async operation has completed).

I think poll-sleeping is a relatively common need in tests, and excessive sleeping makes tests run much slower than they need to.  It's something I've wanted writing GEOM tests, for example.  The proposed mechanism would improve both test performance (stop sleeping when the condition is met) and reliability (rather than some fixed interval, sleep only for as long as it takes to meet the condition, or the timeout expires).

I'm flexible on the exact syntax and semantics of the mechanism, as long as it provides those properties.

ngie-eign commented on atf-sh/atf-check.1:
```diff
@@ -120,6 +120,13 @@ as a shell command line, executing it with the system shell defined by
+(in seconds) expires. If unspecified, the default
+.Ar interval
+(in milliseconds) is 50 ms.
+This can be used to, for example, wait for an expected syslog line.
```
Should this comment be made more general, e.g.,
> This can be used when waiting for the contents of a file to update, for example.

?

ngie-eign commented on atf-sh/atf-check.1:
```diff
@@ -157,6 +164,11 @@ atf_check -s signal:sigsegv my_program
 atf_check -o match:foo -o not-match:bar echo foo baz
+
+# Wait 5 seconds for a line to show up in messages
+logger "testing 123"
```
I don't know if this is a good way to test this functionality in a generic way. I realize that the default behavior for `logger(1)` is to output to `/var/log/messages` on FreeBSD, but this test case relies heavily on defaults.

Would it be better to instead start a background process which asynchronously updates a file, e.g.,

```
test_path=$(mktemp tmp.XXXXXX)
(sleep 10; printf "testing 123\n" > $test_path) &
atf-check -o ignore -e ignore -s exit:0 -r 10 grep "testing 123" $test_path
```

It would probably be a good idea to add multiple variations (`-r` with an interval, without an interval, etc) as well as negative tests too.

ngie-eign replied:
Sure. It's just a good idea to keep in mind the fact that while the fork is against FreeBSD, someone else could take the contents of the fork and try applying them to other OSes.

If this could be cleaned up before merge, cool. Otherwise, we can tackle it at a later date.

cemeyer replied:
Will (try) to address.  Getting subshell and background syntax through roff might be tricky but I'll give it a shot.

ngie-eign commented on atf-sh/atf-check.1:
```diff
@@ -157,6 +164,11 @@ atf_check -s signal:sigsegv my_program
+
+# Wait 5 seconds for a line to show up in messages
+logger "testing 123"
+atf-check -o ignore -e ignore -s exit:0 -r 5 \e
```
Is `\e` here a typo?

ngie-eign commented on atf-sh/atf-check.1:
```diff
@@ -157,6 +164,11 @@ atf_check -s signal:sigsegv my_program
+# Wait 5 seconds for a line to show up in messages
+logger "testing 123"
+atf-check -o ignore -e ignore -s exit:0 -r 5 \e
+    -x grep "testing 123" /var/log/messages
```
This shouldn't require `-x` (spawning a shell).

ngie-eign replied:
This can be cleaned up in a further issue.

cemeyer replied:
Hm, I was going to file an issue for the missing -x and link it here, but it seems like we've disabled Github Issues for our fork, and I don't have permissions to enable them.  Will address the new -x.

ngie-eign commented on atf-sh/atf-check.cpp:
```diff
@@ -162,6 +167,33 @@ class temp_file : public std::ostream {
+    useconds_t res;
+    int rc;
+
+    rc = clock_gettime(CLOCK_MONOTONIC, &ts);
```
Thought: is there a way that `std::chrono::steady_clock::now()` could be used here instead to improve C++'ness of the program: https://en.cppreference.com/w/cpp/chrono/steady_clock/now?

ngie-eign replied:
`clock_gettime` is fine for a first commit. Let's continue :).

cemeyer replied:
Resolving this one for this PR :-).

ngie-eign commented on atf-sh/atf-check.cpp:
```diff
@@ -162,6 +167,33 @@ class temp_file : public std::ostream {
+timo_expired(useconds_t timeout)
+{
+
+    if (get_monotonic_useconds() >= timeout)
```
Some of the `std::chrono::steady_clock` APIs dealing with duration look like a good fit here: https://en.cppreference.com/w/cpp/chrono/steady_clock .

ngie-eign replied:
This can be done later.

ngie-eign commented on atf-sh/atf-check.cpp:
```diff
@@ -306,6 +338,54 @@ parse_output_check_arg(const std::string& arg)
+    // There is no reason this couldn't be a non-integer number of seconds,
+    // this was just easy to do for now.
+    errno = 0;
+    l = strtol(timo_str.c_str(), &end, 10);
```
I'm a bit curious and concerned about overflow issues here with longs vs larger data types. Should this use `strtoll` instead (there might also be a different C++ API that can be used here too)?

cemeyer replied:
It's good to be concerned about this stuff.  In particular, the `strtol` API is bad and difficult to use correctly.  The code is careful not to actually overflow when presented with large input, and instead throw a usage errors.

As far as accepting inputs that exceed the range of `long` — I am comfortable with saying longer timeouts are out of scope for this command.  A 32-bit `long` can represent ~2 billion seconds, or ~63 years, and `long` must be at least 32 bits per the C11 standard.  I anticipate most timeouts to be on the order of `< 60` seconds.

Re: C++ APIs, probably!  There are C++ APIs for everything.  I am broadly unfamiliar with them.  :-)

ngie-eign replied:
This all sounds like good reasoning. `strtol` is ok for now as it works and error checking is added around the return result.

I'll file some follow up issues for investigating the C++ APIs.

cemeyer replied:
Thanks!

ngie-eign commented on atf-sh/atf-check.cpp:
```diff
@@ -306,6 +338,54 @@ parse_output_check_arg(const std::string& arg)
+        throw atf::application::usage_error("Timeout must be a number");
+
+    *m_timo = get_monotonic_useconds() + (l * seconds_in_useconds);
+    *m_interval = 50 * mseconds_in_useconds;
```
Why `50`?

cemeyer replied:
It is arbitrary.  There is no objectively correct default here.

A shorter poll means less time wasted between the wakeup condition presenting and test completion.

A longer poll means less CPU overhead for the polling itself.

50ms was chosen to be long enough that embedded platforms (small, but big enough to run C++ tests!) don't spend significant CPU polling.  It's short enough that it's still vastly more precise than all of the hardcoded test sleeps (and even sleep-poll loops) I've encountered in real tests.

If you'd like a different number, I think a factor of two in either direction could be broadly reasonable.  Or perhaps this deserves a comment mentioning basically the above?  Let me know what action you'd like taken! :-)

ngie-eign replied:
> Or perhaps this deserves a comment mentioning basically the above? Let me know what action you'd like taken! 

A short comment would be great ;). Thanks!

cemeyer replied:
Will do!

ngie-eign commented on atf-sh/atf-check.cpp:
```diff
@@ -162,6 +167,33 @@ class temp_file : public std::ostream {
+    rc = clock_gettime(CLOCK_MONOTONIC, &ts);
+    if (rc != 0)
+        throw std::runtime_error("clock_gettime: " +
+            std::string(strerror(errno)));
```
This can be done later. There are potentially more reentrant/MT issues here and elsewhere that might need to be tackled.

cemeyer replied:
Cool.  Marking resolved as well.

asomers commented:
I don't have strong opinions about it.

cemeyer commented:
Please take a look at the 2nd commit and see if it addresses the concerns raised adequately, and I can squash and force-push something to merge.  Thanks!

cemeyer commented:
@ngie-eign , any chance you can look at the update and see if it addresses your concerns adequately?  If so, I'd like to squash and merge this.  Thanks!
